src/levanter/compat/fsspec_safetensor.py

```diff
@@ -0,0 +1,290 @@
+# Copyright 2025 The Levanter Authors
+# SPDX-License-Identifier: Apache-2.0
+
+import asyncio
+import json
+import logging
+import os
+import struct
+from dataclasses import dataclass
+from typing import Callable, Dict, List, Optional, Tuple, Iterable
+
+import fsspec
+import humanfriendly
+import jax
+import jax.numpy as jnp
+import numpy as np
+import tqdm_loggable.auto as tqdm
+from fsspec import AbstractFileSystem
+from fsspec.asyn import AsyncFileSystem
+
+logger = logging.getLogger(__name__)
+
+_SAFETENSOR_DTYPE_MAP: Dict[str, np.dtype] = {
+    "F16": np.dtype("float16"),
+    "BF16": np.dtype(jnp.bfloat16),
+    "F32": np.dtype("float32"),
+    "F64": np.dtype("float64"),
+    "I8": np.dtype("int8"),
+    "I16": np.dtype("int16"),
+    "I32": np.dtype("int32"),
+    "I64": np.dtype("int64"),
+    "U8": np.dtype("uint8"),
+    "U16": np.dtype("uint16"),
+    "U32": np.dtype("uint32"),
+    "U64": np.dtype("uint64"),
+    "BOOL": np.dtype("bool"),
+}
+
+
+ShardingFunction = Callable[[Tuple[int, ...]], Optional[jax.sharding.Sharding]]
+
+DEFAULT_CHUNK_SIZE_BYTES = int(os.environ.get("LEVANTER_FSSPEC_CHUNK_BYTES", 2 * 1024**3))
```

Comment:
I guess I'm confused about this part. If we need to read the whole file anyway, then we can just pre-allocate 1 giant buffer and launch a threadpool to fill it up in memory.

If we don't have to read the whole file, this chunk size is way to big.

I suspect you want a chunk size like 64MB to get more parallelism out of GCS though.

Reply:
we don't **really** need to read in the whole tensor on every worker, but every time i try to read just shards on 4 machines it's worse than doing it this way (which is why this code exists in a bit of a frankensttate. I can't quite give up on the dream of reading just the data we need, or something)